usr/src/uts/common/fs/zfs/dmu.c

```diff
@@ -690,10 +758,16 @@ dmu_free_long_range_impl(objset_t *os, dnode_t *dn, uint64_t offset,
 			dmu_tx_abort(tx);
 			return (err);
 		}
-		dnode_free_range(dn, chunk_begin, chunk_end - chunk_begin, tx);
+		mutex_enter(&dp->dp_lock);
+		dp->dp_long_free_dirty_pertxg[dmu_tx_get_txg(tx) & TXG_MASK] +=
+		    chunk_len;
```

Comment:
Why increment this and dp_long_freeing_total in different places?  Seems like it makes it easy for them to not be in sync.  E.g. if we "return (err)" on line 759, we will have incremented dp_long_freeing_total but not dp_long_free_dirty_pertxg[].  IMO, all the more reason to get rid of dp_long_freeing_total.